packages/askar/src/wallet/AskarWallet.ts

```diff
@@ -0,0 +1,745 @@
+    const message = cek.aeadDecrypt({
+      ciphertext: new Uint8Array(Buffer.from(messagePackage.ciphertext as any, 'base64')),
+      nonce: new Uint8Array(Buffer.from(messagePackage.iv as any, 'base64')),
+      tag: new Uint8Array(Buffer.from(messagePackage.tag as any, 'base64')),
+      aad: new TextEncoder().encode(messagePackage.protected),
+    })
+
+    return {
+      plaintextMessage: JsonEncoder.fromBuffer(message),
+      senderKey,
+      recipientKey,
+    }
+  }
+
+  public async generateNonce(): Promise<string> {
+    try {
+      return new TextDecoder().decode(CryptoBox.randomNonce())
+    } catch (error) {
+      if (!isError(error)) {
+        throw new AriesFrameworkError('Attempted to throw error, but it was not of type Error', { cause: error })
+      }
+      throw new WalletError('Error generating nonce', { cause: error })
+    }
+  }
+
+  public async generateWalletKey() {
+    try {
+      return Store.generateRawKey()
+    } catch (error) {
+      throw new WalletError('Error generating wallet key', { cause: error })
+    }
+  }
+
+  private async getAskarWalletConfig(walletConfig: WalletConfig) {
+    const { uri, path } = uriFromWalletConfig(walletConfig, this.fileSystem.basePath)
+
+    // Make sure path exists before creating the wallet
+    if (path) {
+      await this.fileSystem.createDirectory(path)
+    }
+
+    return {
+      uri,
+      profile: walletConfig.id,
+      // FIXME: Default derivation method should be set somewhere in either agent config or some constants
+      keyMethod: keyDerivationMethodToStoreKeyMethod(walletConfig.keyDerivationMethod) ?? StoreKeyMethod.Raw,
+      passKey: walletConfig.key,
+    }
+  }
+
+  private async retrieveKeyPair(publicKeyBase58: string): Promise<KeyPair> {
+    try {
+      const entryObject = await this.session.fetch({ category: 'KeyPairRecord', name: `key-${publicKeyBase58}` })
+
+      if (entryObject?.value) {
+        return JsonEncoder.fromString(entryObject?.value as string) as KeyPair
+      } else {
+        throw new WalletError(`No content found for record with public key: ${publicKeyBase58}`)
+      }
+    } catch (error) {
+      if (
+        isAskarError(error) &&
+        (error.code === askarErrors.NotFound ||
+          // FIXME: this is current output from askar wrapper but does not describe specifically a not found scenario
+          error.message === 'Received null pointer. The native library could not find the value.')
+      ) {
+        throw new RecordNotFoundError(`KeyPairRecord not found for public key: ${publicKeyBase58}.`, {
+          recordType: 'KeyPairRecord',
+          cause: error,
+        })
+      }
+      throw new WalletError('Error retrieving KeyPair record', { cause: error })
+    }
+  }
+
+  private async storeKeyPair(keyPair: KeyPair): Promise<void> {
+    try {
+      await this.session.insert({
+        category: 'KeyPairRecord',
+        name: `key-${keyPair.publicKeyBase58}`,
```

Comment:
Is this the same category / id as the indy-sdk? That'll be important for the migration script

Reply:
Yes. And as Askar is supporting natively more algorithms than Indy, the migration script must make sure to write all those keys natively supported into the specific Key Storage, so it will not be a simply "copy-paste".